### ISSUE TITLE
[CI] JwtRealmAuthenticateTests testJwkSetUpdates failing

### ISSUE THREAD

**Issue by gmarouli** [elastic/elasticsearch#90467]
It looks like the test started failing on the 23rd on `8.4` and `8.5`: https://gradle-enterprise.elastic.co/scans/tests?search.startTimeMax=1664376109074&search.startTimeMin=1663711200000&search.timeZoneId=Europe/Amsterdam&tests.container=org.elasticsearch.xpack.security.authc.jwt.JwtRealmAuthenticateTests&tests.test=testJwkSetUpdates#

**Build scan:**
https://gradle-enterprise.elastic.co/s/mh6xbsqlllxki/tests/:x-pack:plugin:security:test/org.elasticsearch.xpack.security.authc.jwt.JwtRealmAuthenticateTests/testJwkSetUpdates

**Reproduction line:**
`./gradlew ':x-pack:plugin:security:test' --tests "org.elasticsearch.xpack.security.authc.jwt.JwtRealmAuthenticateTests.testJwkSetUpdates" -Dtests.seed=95E5E925085855C9 -Dtests.locale=es-HN -Dtests.timezone=America/Dawson -Druntime.java=17`

**Applicable branches:**
8.5, 8.4

**Reproduces locally?:**
No

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.xpack.security.authc.jwt.JwtRealmAuthenticateTests&tests.test=testJwkSetUpdates

**Failure excerpt:**
```
java.lang.AssertionError: Expected realm CONTINUE status
Expected: is not "realm_iss1_4421"
     but: was "realm_iss1_4421"

  at __randomizedtesting.SeedInfo.seed([95E5E925085855C9:A73A48ACF157045D]:0)
  at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
  at org.junit.Assert.assertThat(Assert.java:956)
  at org.elasticsearch.xpack.security.authc.jwt.JwtRealmTestCase.doMultipleAuthcAuthzAndVerifySuccess(JwtRealmTestCase.java:526)
  at org.elasticsearch.xpack.security.authc.jwt.JwtRealmAuthenticateTests.testJwkSetUpdates(JwtRealmAuthenticateTests.java:241)
  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(NativeMethodAccessorImpl.java:-2)
  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
  at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  at java.lang.reflect.Method.invoke(Method.java:568)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:44)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:833)

```

**Comment by ywangd** [comment]
It might be related to #89509. I believe Justin is working on reproduction. I can also take a look at this one from the perspective of racing. Self-assigned.

**Comment by cbuescher** [comment]
Another one today on the 8.4 branch https://gradle-enterprise.elastic.co/s/iec5kcsddvjvc

**Comment by cbuescher** [comment]
Another one on 8.5 branch, but the error looks slightly different. Still adding here since it looks related:
https://gradle-enterprise.elastic.co/s/ewmjvx727737i

```
java.lang.AssertionError: 
Expected: is <false>
     but: was <true>
	at __randomizedtesting.SeedInfo.seed([374EF16B6DC63043:59150E294C961D7]:0)
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
	at org.junit.Assert.assertThat(Assert.java:956)
	at org.junit.Assert.assertThat(Assert.java:923)
	at org.elasticsearch.xpack.security.authc.jwt.JwtRealmTestCase.verifyAuthenticateFailureHelper(JwtRealmTestCase.java:121)
	at org.elasticsearch.xpack.security.authc.jwt.JwtRealmAuthenticateTests.testJwkSetUpdates(JwtRealmAuthenticateTests.java:227)
```

**Comment by ywangd** [comment]
I spent quite sometime today trying to figure out why the test is failing but didn't make a lot progress. The test in question is rather hard to follow. Long term we might want to split it into a few smaller ones. For now, I can see the PKC JWKs reloading is not triggered properly. 

The failure message says (both JWT uses PKC algorithm)
```
1. [[I/O dispatcher 97]] Usable PKC: JWKs=[8] algorithms=[RS512,PS256,ES256,ES384,ES512,PS384,ES384,ES384] sha256=[9f0bccf33c63441f07144bb0e6852d831d9b6cde737dd9b3b72ec3a64ab69829]
2. [[I/O dispatcher 97]] Usable PKC: JWKs=[8] algorithms=[RS512,PS256,ES256,ES384,ES512,PS384,ES384,ES384] sha256=[c3191bf8b73c7c4ebb39fc8c9fdf1a477b82fa786c317291fa4d31b95c273d5b]
3. [[I/O dispatcher 97]] Usable PKC: JWKs=[8] algorithms=[RS512,PS256,ES256,ES384,ES512,PS384,ES384,ES384] sha256=[c3191bf8b73c7c4ebb39fc8c9fdf1a477b82fa786c317291fa4d31b95c273d5b]
4. [[I/O dispatcher 97]] Usable PKC: JWKs=[0] algorithms=[] sha256=[9638d11914735d0072002dd4d67c734a94ee7a0cd845e6dcbc984ee7581860a7]
5. [[I/O dispatcher 97]] Usable PKC: JWKs=[0] algorithms=[] sha256=[9638d11914735d0072002dd4d67c734a94ee7a0cd845e6dcbc984ee7581860a7]
```

Where each line should correspond to:
1. [Initial realm creation](https://github.com/elastic/elasticsearch/blob/d1f2b5d79ddfae4e63777960f1406931abca80c5/x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealmAuthenticateTests.java#L101) sha: 9f0bccf33c63441f07144bb0e6852d831d9b6cde737dd9b3b72ec3a64ab69829
2. Triggerred by [authenticating with JWT2 the first time](https://github.com/elastic/elasticsearch/blob/d1f2b5d79ddfae4e63777960f1406931abca80c5/x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealmAuthenticateTests.java#L179-L185) sha: c3191bf8b73c7c4ebb39fc8c9fdf1a477b82fa786c317291fa4d31b95c273d5b
3. Triggered by [re-authenticating with JWT1](https://github.com/elastic/elasticsearch/blob/d1f2b5d79ddfae4e63777960f1406931abca80c5/x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealmAuthenticateTests.java#L192-L196) sha: c3191bf8b73c7c4ebb39fc8c9fdf1a477b82fa786c317291fa4d31b95c273d5b
4. Triggered by [re-re-authenticating with JWT1](https://github.com/elastic/elasticsearch/blob/d1f2b5d79ddfae4e63777960f1406931abca80c5/x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealmAuthenticateTests.java#L213-L217), the JWT set is empty this time. sha: 9638d11914735d0072002dd4d67c734a94ee7a0cd845e6dcbc984ee7581860a7
5. Triggered by [re-authenticating with JWT2](https://github.com/elastic/elasticsearch/blob/d1f2b5d79ddfae4e63777960f1406931abca80c5/x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealmAuthenticateTests.java#L224-L228). Still empty JWK set sha: 9638d11914735d0072002dd4d67c734a94ee7a0cd845e6dcbc984ee7581860a7
6. **This is where it failed**. It should be triggered by [re-re-authenticating with JWT2](https://github.com/elastic/elasticsearch/blob/d1f2b5d79ddfae4e63777960f1406931abca80c5/x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealmAuthenticateTests.java#L240-L244). But the authentication simply failed without a reload. 

It is not clear to me why the last reload did not happen. There could be an exception involved but the test does not log it. I think the next step is to add this missing logging information and perhaps a few more around other critical regions.

**Comment by justincr-elastic** [comment]
@ywangd a design change that came up in team discussion was only allowing each realm to have PKC or HMAC. In other words, if you want both types of keys from the same iss+aud, create two realm instances.

I think that would simplify the realm code, and really simplify the JWK update tests in this issue. It would make tracing intermittent CI failures much, much easier.

Is forcing PKC and HMAC into different JWT realm instances something you included on your wish list for JWT realm improvements? It is technically a breaking change, but the feature is technically still in beta, and customers would not lose PKC+HMAC functionality. It might have enough benefits to take it to the breaking change committee if needed.

**Comment by ywangd** [comment]
> Is forcing PKC and HMAC into different JWT realm instances something you included on your wish list for JWT realm improvements?

Short answer is No. 

I initially thought was tempted to do that as well. But later I realised that the oidc relam actually does [support both HMAC and PKC within a single relam](https://github.com/elastic/elasticsearch/blob/5596c27dca6e461a253edfacec1da4d72967c017/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectAuthenticator.java#L772-L776). If we use oidc as an analogy, I think JWT realm should actually support both with a single realm as well. That said, client secret is used as a password by the oidc realm which is not something we do for the JWT realm. So maybe the analogy is not entirely accurate. I think we should keep this as an open discussion. But I don't want to necessarily tie it to the current JWT work.
